### PR TITLE
Nexus (also Digoo?) sensors

### PR DESCRIPTION
**Pull request by ligius-**
Add support for Nexus sensors

### CODE REVIEW

**Comment by CurlyMoo** [comment]
PR's against the master branch are not allowed and the PR shows merge conflicts against the staging branch. Can you fix those?

File: libs/pilight/protocols/433.92/nexus.c [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            binary[i++] = 1;
+        } else if (isValidPulse(nexus->raw[x], ZERO_P)) {
+            binary[i++] = 0;
+        } else {
+            return;
+        }
+    }
+
+    if (binary[9] != 0) {
+        // logprintf(LOG_DEBUG, "nexus: Bit 10 not zero");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: Can you remove these commected `logprintf` statements?

File: libs/pilight/protocols/433.92/nexus.c [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    battery = binary[8] ? 100 : 0;
+    channel = binToDecRev(binary, 10, 11);
+    temperature = (double)binToSignedRev(binary, 12, 23);
+    humidity = (double)binToDecRev(binary, 28, 35);
+
+    temperature /= 10;
+
+    struct settings_t *tmp = settings;
+    while (tmp) {
+        if (fabs(tmp->id - id) < EPSILON) {
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: The `id` is an integer, why is handle it as a double?

File: libs/pilight/protocols/433.92/nexus.c [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                    id = jchild1->number_;
+                }
+                jchild1 = jchild1->next;
+            }
+            jchild = jchild->next;
+        }
+
+        struct settings_t *tmp = settings;
+        while (tmp) {
+            if (fabs(tmp->id - id) < EPSILON) {
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: Likewise

File: libs/pilight/protocols/433.92/nexus.c [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        struct JsonNode *jchild1 = NULL;
+        double id = -1;
+        int match = 0;
+
+        jchild = json_first_child(jid);
+        while (jchild) {
+            jchild1 = json_first_child(jchild);
+            while (jchild1) {
+                if (strcmp(jchild1->key, "id") == 0) {
+                    id = jchild1->number_;
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: You fixed the if statement below, but that requires a cast to int here.

File: libs/pilight/protocols/433.92/nexus.c [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+/**
+ * Validate whether a raw pulse length matches a known type
+ */
+static bool isValidPulse(uint16_t pulseLength, enum PulseType measureAgainst) {
+    return (pulseLength > (measureAgainst - PULSE_TOLERANCE)) && (pulseLength < (measureAgainst + PULSE_TOLERANCE));
+}
+
+static int validate(void) {
+    if (nexus->rawlen >= RAW_LENGTH && isValidPulse(nexus->raw[0], START_P) && isValidPulse(nexus->raw[RAW_LENGTH - 2], START_P)) {
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: Small comment, but can you follow pilight style guide regarding the conditions.
`if(` instead of `if (`

Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: I just used CLang with the Google style, are there any other formatting errors? I will fix all of them

File: libs/pilight/protocols/433.92/nexus.c [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    if (binary[9] != 0) {
+        return;
+    }
+    if (binary[24] != 1 || binary[25] != 1 || binary[26] != 1 || binary[27] != 1) {
+        return;
+    }
+
+    // decode bits into data
+    id = binToDecRev(binary, 0, 7);
+    battery = binary[8] ? 100 : 0;
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: What's up with that battery `100` value?

Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: There's no actual reporting of the battery value, just of a low battery. That's how the other decoders are handling this.

Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: Why not a simple true / false as in 1 / 0?

File: libs/pilight/protocols/433.92/nexus.c [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    channel = binToDecRev(binary, 10, 11);
+    temperature = (double)binToSignedRev(binary, 12, 23);
+    humidity = (double)binToDecRev(binary, 28, 35);
+
+    temperature /= 10;
+
+    struct settings_t *tmp = settings;
+    while (tmp) {
+        if (tmp->id == id) {
+            humi_offset = tmp->humi;
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: You can just do the math for temperature and humidity here directly. No need to store them into temp. variables.

File: libs/pilight/protocols/433.92/nexus.c [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    temperature += temp_offset;
+    humidity += humi_offset;
+
+    nexus->message = json_mkobject();
+    json_append_member(nexus->message, "id", json_mknumber(id, 0));
+    json_append_member(nexus->message, "channel", json_mknumber(channel, 0));
+    json_append_member(nexus->message, "battery", json_mknumber(battery, 0));
+    json_append_member(nexus->message, "temperature", json_mknumber(temperature, 1));
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: This `1` doesn't take the `temperature-decimals` into account.

File: libs/pilight/protocols/433.92/nexus.c (hunk 7 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: I just copied what the others were doing. Is this offset something that can be set from the settings/UI? I have no idea what it does.

File: libs/pilight/protocols/433.92/nexus.c (hunk 5 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: Not errors, just preferences.

File: libs/pilight/protocols/433.92/nexus.c (hunk 7 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: I know, but we learn :smile:

I can be done more efficiently. These values can be set the config indeed.

File: libs/pilight/protocols/433.92/nexus.c (hunk 8 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: Will remove that option

File: libs/pilight/protocols/433.92/nexus.c (hunk 5 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: Are there any other options to CLang? I have these: { BasedOnStyle: Google, IndentWidth: 4, ColumnLimit: 0}

File: libs/pilight/protocols/433.92/nexus.c (hunk 8 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: Why not make the decimal variable here?

File: libs/pilight/protocols/433.92/nexus.c (hunk 5 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: I don't now, i just use `notepad++` and manually type my code.

Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: I added SpaceBeforeParens: Never to my VSCode configuration, seems to the the job. Perhaps this should be documented into the "Contributing" section:
{ BasedOnStyle: Google, IndentWidth: 4, ColumnLimit: 0, SpaceBeforeParens: Never }

Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: I see a second PR comming :wink:

File: libs/pilight/protocols/433.92/nexus.c (hunk 6 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: Is there any example that already does this? I saw the generic_weather one has 0/1 on the battery, but what does that mean? That there is a battery installed? That it has voltage? That it's low?
The parameters say "set the battery level", so 1 means full?

File: libs/pilight/protocols/433.92/nexus.c (hunk 8 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: I don't know if there is any benefit to that. The sensor reports 1 decimal always. What can I add to make it like you say? I don't know how to program it.

File: libs/pilight/protocols/433.92/nexus.c (hunk 6 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: Good point, but i have found no weather station protocol that actually reported a non-binary value. Until we do, the GUI just parses a binary value: empty / not-empty.

File: libs/pilight/protocols/433.92/nexus.c (hunk 8 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: But people can choose to round the single decimal to no decimals.  Just store the value in the settings struct and replace the `1` in the json code with that variable.

**Comment by ligius-** [comment]
It passes the smoke test (compilation + pilight-daemon) but I don't have my sensor connected to the testing Pi anymore, so I hope everything still works. I haven't played with any of the settings. In fact, yesterday was the first time I checked out and installed pilight and this is my first PR on GitHub. Plus, I don't write C code, except as light hobby :)

File: libs/pilight/protocols/433.92/nexus.c (hunk 6 above)
Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: So 0 is empty?

Comment (ligius-) [reply libs/pilight/protocols/433.92/nexus.c]: So 0 is empty?

File: libs/pilight/protocols/433.92/nexus.c [hunk 9]
```diff
@@ -121,43 +121,32 @@ static void parseCode(void) {
     humidity = (double)binToDecRev(binary, 28, 35);
 
     temperature /= 10;
-
-    struct settings_t *tmp = settings;
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: I meant this:
```c
    struct settings_t *tmp = settings;
    while (tmp) {
        if (tmp->id == id) {
            temperature += tmp->temp;
            humidity += tmp->humi;
            break;
        }
        tmp = tmp->next;
    }
```

File: libs/pilight/protocols/433.92/nexus.c [hunk 10]
```diff
@@ -121,43 +121,32 @@ static void parseCode(void) {
...
-    humidity += humi_offset;
+    temperature += settings->temperature_offset;
+    humidity += settings->humidity_offset;
 
     nexus->message = json_mkobject();
     json_append_member(nexus->message, "id", json_mknumber(id, 0));
     json_append_member(nexus->message, "channel", json_mknumber(channel, 0));
     json_append_member(nexus->message, "battery", json_mknumber(battery, 0));
-    json_append_member(nexus->message, "temperature", json_mknumber(temperature, 1));
+    json_append_member(nexus->message, "temperature", json_mknumber(temperature, settings->temperature_decimals));
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/nexus.c]: This should be stored in a temporary value from within the while loop above.

File: libs/pilight/protocols/433.92/nexus.c (hunk 6 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/nexus.c]: Yes

**Comment by ligius-** [comment]
I tested the changes and they work fine after my last commit, they were crashing with the commit before it.

For reference, here are two JSON messages from two sensors, one of them has a low battery (=0):
```
{"message":{"id":158,"channel":0,"battery":0,"temperature":23.2,"humidity":61},"protocol":"nexus","length":"158","repeats":2,"status":2}
{"message":{"id":98,"channel":1,"battery":1,"temperature":22.9,"humidity":62},"protocol":"nexus","length":"98","repeats":2,"status":2}
```

The code before creating this PR has been running for a few months now in an OpenMQTTGateway custom build which includes ESPiLight which includes pilight.

**Comment by CurlyMoo** [comment]
Can you do the similar PR on the rewrite branch including some unittests?

**Comment by ligius-** [comment]
I just looked at that branch, it doesn't look like many changes are required, just adding the tests. I can start working on that, but no promise on the timeline.

Is merging into master depending on the rewrite branch? Or is that a parallel effort and not really related?

**Comment by CurlyMoo** [comment]
> I just looked at that branch, it doesn't look like many changes are required, just adding the tests. I can start working on that, but no promise on the timeline.

Indeed

> Is merging into master depending on the rewrite branch? Or is that a parallel effort and not really related?

Things never get merged directly into master. Always first in staging. Mergin is indeed dependent on the rewrite PR.

**Comment by ligius-** [comment]
Are there any other findings that do you see coming up, related to this PR?

Is the 'rewrite' branch branched from staging, i.e., does 'staging' get merged into it whenever a PR is approved here? Or are the PRs there cherry-picked from this branch?

Does merging to 'master' (from 'staging') require that similar code is committed to 'staging' and 'rewrite'?

Obviously I can't add tests to 'rewrite' until this code is already there. Having more findings here ('staging') means dual-maintenance if I would already have a PR open in 'rewrite'.

**Comment by CurlyMoo** [comment]
>Are there any other findings that do you see coming up, related to this PR?

Loads, but not anything i'm going to bother you with :smile:

>  Or are the PRs there cherry-picked from this branch?

The PR's from rewrite get cherry picked into staging.

> Does merging to 'master' (from 'staging') require that similar code is committed to 'staging' and 'rewrite'?

Ja

> Obviously I can't add tests to 'rewrite' until this code is already there. Having more findings here ('staging') means dual-maintenance if I would already have a PR open in 'rewrite'.

Yes, once merged, that's my problem.

**Comment by melyux** [comment]
I've tried this code on my Pilight and it works great.

**Comment by CurlyMoo** [comment]
Can you also add a manual page?

**Comment by melyux** [comment]
Is it possible to allow transmission as well? Right now I’m transmitting the Nexus protocol as raw code on Pilight. I don’t think I’ve seen any supported temperature sensors supporting transmission yet though.

**Comment by ligius-** [comment]
> Is it possible to allow transmission as well? Right now I’m transmitting the Nexus protocol as raw code on Pilight. I don’t think I’ve seen any supported temperature sensors supporting transmission yet though.

I don't see why not, it should be fairly easy to do. But I wouldn't want to add new features until I see both pull requests merged. Also, not sure how much sense this makes from pilight's point of view, I haven't seen any transmission done on the sensors.

**Comment by ligius-** [comment]
> Can you also add a manual page?

I added it to this PR/branch, haven't seen a documentation folder on the rewrite branch.

**Comment by CurlyMoo** [comment]
Did you also compile the documentation?

**Comment by ligius-** [comment]
> Did you also compile the documentation?

Not yet, do you see any potential issues? Is the documentation uploaded with the PR or generated elsewhere?

**Comment by CurlyMoo** [comment]
> Not yet, do you see any potential issues? Is the documentation uploaded with the PR or generated elsewhere?

I will check it tomorrow. The https://manual.pilight.org is automatically regenerated once a new commit has been done to the staging branch.

**Comment by ligius-** [comment]
I installed the needed dependencies and generated the documentation (dirhtml) locally. Looks fine, but you can double-check:

image

**Comment by CurlyMoo** [comment]
There is still that bogus `"` in the supported brands where: `“Nexus Weather Stations`

File: docs/source/protocols/433.92/weather/_nexus.rst [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+*None*
+
+.. rubric:: Config
+
+.. code-block:: json
+   :linenos:
+
+   {
+     "devices": {
```
Comment (CurlyMoo) [inline docs/source/protocols/433.92/weather/_nexus.rst]: This config example fails on the channel because it's not defined an ID value.

**Comment by CurlyMoo** [comment]
Did you validate the new config yourself?

**Comment by ligius-** [comment]
> Did you validate the new config yourself?

Yes, I started pilight-daemon and saw the sensor placeholder there. There were no errors (anymore) in pilight.err. The rewrite branch has also been updated.

**Pull request merged**